zcash_extensions/src/transparent/demo.rs

```diff
@@ -273,15 +279,100 @@ impl<C: Context> Extension<C> for Program {
     }
 }
 
+fn builder_hashes(preimage_1: &[u8; 32], preimage_2: &[u8; 32]) -> ([u8; 32], [u8; 32]) {
+    let hash_2 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(Params::new().hash_length(32).hash(preimage_2).as_bytes());
+        hash
+    };
+
+    let hash_1 = {
+        let mut hash = [0; 32];
+        hash.copy_from_slice(
+            Params::new()
+                .hash_length(32)
+                .to_state()
+                .update(preimage_1)
+                .update(&hash_2)
+                .finalize()
+                .as_bytes(),
+        );
+        hash
+    };
+
+    (hash_1, hash_2)
+}
+
+pub struct DemoBuilder<'a, B> {
+    txn_builder: &'a mut B,
+    extension_id: usize,
+}
+
+impl<'a, B: ExtensionTxBuilder<'a>> DemoBuilder<'a, B> {
+    pub fn demo_open(
+        &mut self,
+        value: Amount,
+        preimage_1: [u8; 32],
+        preimage_2: [u8; 32],
+    ) -> Result<(), B::BuildError> {
+        let (hash_1, _) = builder_hashes(&preimage_1, &preimage_2);
+
+        // Call through to the generic builder.
+        self.txn_builder
+            .add_tze_output(self.extension_id, value, &Precondition::open(hash_1))
+    }
+
+    pub fn demo_transfer_to_close(
+        &mut self,
+        prevout: (OutPoint, TzeOut),
+        transfer_amount: Amount,
+        preimage_1: [u8; 32],
+        preimage_2: [u8; 32],
+    ) -> Result<(), B::BuildError> {
+        let (_, hash_2) = builder_hashes(&preimage_1, &preimage_2);
+
+        // should we eagerly validate the relationship between prevout.1 and preimage_1?
+        self.txn_builder
+            .add_tze_input(self.extension_id, prevout, move |_| {
+                Ok(Witness::open(preimage_1))
+            })?;
+
+        self.txn_builder.add_tze_output(
+            self.extension_id,
+            transfer_amount, // can this be > prevout.1.value?
+            &Precondition::close(hash_2),
+        )
+    }
+
+    pub fn demo_close(
+        &mut self,
+        prevout: (OutPoint, TzeOut),
+        preimage: [u8; 32],
+    ) -> Result<(), B::BuildError> {
+        let hash_2 = {
+            let mut hash = [0; 32];
+            hash.copy_from_slice(Params::new().hash_length(32).hash(&preimage).as_bytes());
+            hash
+        };
+
+        self.txn_builder
+            .add_tze_input(self.extension_id, prevout, move |_| {
+                Ok(Witness::close(hash_2))
```

Comment:
```suggestion
                Ok(Witness::close(preimage))
```

Also rename it to `preimage_2` for consistency.

Reply:
preimage_2 is the argument that is used to construct the hash; isn't hash_2 correct here? We don't have a `preimage` argument in this scope.

Reply:
The hashes go in precommitments; the preimages go in witnesses.